File: 498-diagonal-traverse/498-diagonal-traverse.cs
public class Solution {
    public int[] FindDiagonalOrder(int[][] mat) {
        int m = mat.Length, n = mat[0].Length;
        
        List<int> result = new List<int>();
        int index = 0;
        
        //down the rows
        bool reverse = false;
        for(int i = 0; i < m; i++){
            int row = i, col = 0;
            List<int> res = new List<int>();
            while(row >= 0 && col < n){
                res.Add(mat[row][col]);
                if(row-1 >= 0 && col+1 < n){
                    row--;
                    col++;
                }
                else
                    break;
            }
            
            if(reverse){
                res.Reverse();
                reverse = false;
            }
            else{
                reverse = true;
            }
            
            result.AddRange(res);
        }
        
        //Console.WriteLine("done");
        
        //reverse = false;
        for(int j = 1; j < n; j++){
            int row = m-1, col = j;
            List<int> res = new List<int>();
            while(row >= 0 && col < n){
                res.Add(mat[row][col]);
                if(row-1 >= 0 && col+1 < n){
                    row--;
                    col++;
                }
                else
                    break;
            }
            
            if(reverse){
                res.Reverse();
                reverse = false;
            }
            else{
                reverse = true;
            }
            
            result.AddRange(res);
        }
        
        return result.ToArray();
    }
}

File: 498-diagonal-traverse/498-diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] result = new int[m*n];
        
        int total = m*n;
        int row = 0, col = 0;
        int count = 0;
        int index= 0;
        while(count < total){
            while(row >= 0 && col < n){
                if(count == total)
                    break;
                
                result[index++] = mat[row][col];
                count++;
                
                row--;
                col++;
            }
            
            if(col == n){
                col = n-1;
                row += 2;
            }
            else if(row < 0){
                row = 0;
            }
            
            while(col >= 0 && row < m){
                if(count == total)
                    break;
                
                result[index++] = mat[row][col];
                count++;
                row++;
                col--;
            }
            
            if(row == m){
                row = m-1;
                col += 2;
            }
            else if(col < 0){
                col = 0;
            }
        }
        
        return result;
    }
}